modules/odr_api/odr_api/api/endpoints/auth.py

```diff
@@ -46,7 +68,139 @@ def logout(db=Depends(get_db), session_cookie: str = Depends(get_session_cookie)
+        raise HTTPException(status_code=401, detail="Different provider already used")
+
+    response = RedirectResponse(url=f"{request.base_url}{settings.OAUTH2_REDIRECT_PATH}")
+    response.set_cookie(key="session", value=session.id, httponly=True)
+
+    return response
+
+
+# Google SSO
+google_sso = GoogleSSO(
+    client_id=settings.GOOGLE_CLIENT_ID,
+    client_secret=settings.GOOGLE_CLIENT_SECRET,
+    redirect_uri="http://localhost:31100/api/v1/auth/google/callback",
+)
+
+
+@router.get("/auth/google/login")
+async def google_login(request: Request):
+    with google_sso:
+        return await google_sso.get_login_redirect(
+            redirect_uri=request.url_for("google_callback")
+        )
+
+
+@router.get("/auth/google/callback")
+async def google_callback(request: Request, db: Session = Depends(get_db)):
+    with google_sso:
+        try:
+            user = await google_sso.verify_and_process(request)
+        except Exception:
+            raise HTTPException(status_code=401, detail="Failed to login with Google")
+        return ouath2_login_and_signup(request, user, db)
+
+
+# Github SSO
+github_sso = GithubSSO(
+    client_id=settings.GITHUB_CLIENT_ID,
+    client_secret=settings.GITHUB_CLIENT_SECRET,
+)
+
+
+@router.get("/auth/github/login", description="Login with Github")
+async def github_login(request: Request):
+    with github_sso:
+        return await github_sso.get_login_redirect(
+            redirect_uri=request.url_for("github_callback")
+        )
+
+
+@router.get("/auth/github/callback")
+async def github_callback(request: Request, db: Session = Depends(get_db)):
+    with github_sso:
+        try:
+            user = await github_sso.verify_and_process(request)
+        except Exception:
+            raise HTTPException(status_code=401, detail="Failed to login with Github")
+        return ouath2_login_and_signup(request, user, db)
+
+
+# Discord SSO
+def convert_discord_openid(
+    response: Dict[str, Any], _client: Union[AsyncClient, None]
+) -> OpenID:
+    """Convert user information returned by OIDC"""
+
+    if "user" not in response:
+        raise HTTPException(status_code=400, detail="Invalid response from Discord")
+
+    user = response["user"]
+
+    return OpenID(
+        id=user.get("id"),
+        name=user.get("username"),
+        email=user.get("email"),
+        picture=user.get("avatar"),
+        provider="discord",
+    )
+
+
+discovery_document: DiscoveryDocument = {
+    "authorization_endpoint": "https://discord.com/oauth2/authorize",
+    "token_endpoint": "https://discord.com/api/oauth2/token",
+    "userinfo_endpoint": "https://discord.com/api/oauth2/@me",
+}
+
+
+DiscordSSO = create_provider(
+    name="discord",
+    default_scope=["identify", "email"],
+    discovery_document=discovery_document,
+    response_convertor=convert_discord_openid,
+)
+
+
+discord_sso = DiscordSSO(
+    client_id=settings.DISCORD_CLIENT_ID,
+    client_secret=settings.DISCORD_CLIENT_SECRET,
+)
```

Comment:
Does Discord SSO not need a redirect uri to api/v1/auth/discord/callback similar to Google?

Reply:
The github configuration was the wrong one.
Redirect uri is calculated automatically using this method. So this property is not needed.
```
        return await discord_sso.get_login_redirect(
            redirect_uri=request.url_for("discord_callback")
        )
```